google/cloud/spanner/result_set.h

```diff
@@ -0,0 +1,111 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_SPANNER_RESULT_SET_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_SPANNER_RESULT_SET_H_
+
+#include "google/cloud/spanner/internal/time.h"
+#include "google/cloud/spanner/row_parser.h"
+#include "google/cloud/spanner/timestamp.h"
+#include "google/cloud/optional.h"
+#include <google/spanner/v1/spanner.pb.h>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+
+namespace internal {
+class ResultSetValueSource {
```

Comment:
(optional) This is more than a "ValueSource", so maybe removing ValueSource from the name would be good. Other names that might make sense "ResultSetImpl" (boring, but a classic), "ResultSetSource" (kinda meaningful).